Revert to status branch

Darcy3000s opened the pull request.

Darcy3000s committed:
Add a button in the right click context menu on assets which allows you to "Revert to status branch" (Rollback to Origin) - notably this doesn't actually perform any reverts, it just pulls the latest version of that file from main into your branch

Darcy3000s committed:
Diff files before attempting to sync them from another branch so we only sync files which are actually different - bail if we try to sync a file that we don't have to ability to checkout (It's locked by somebody else or changed in a branch other than status branch). Don't try to checkout files which are already checked out

Darcy3000s committed:
Change icon for Status Revert to use Revert icon

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2532,6 +2535,80 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+	TArray<FString> DiffResults;
+	TArray<FString> ErrorMessages;
+	TArray<FString> DiffParametersLog{ TEXT("--pretty="), TEXT("--name-only"), FString::Printf(TEXT("HEAD..%s"), *BranchName), TEXT(""), TEXT("--") };
+	const bool bResultDiff = RunCommand(TEXT("diff"), InPathToGitBinary, InRepositoryRoot, DiffParametersLog, FilesToSync, DiffResults, ErrorMessages);
```
FilesToSync is the wrong file path type. We need to make sure these are in the Unreal file paths

Damon3000s replied:
We DON'T need to make sure these are in the Unreal path types. As the new comments in the code mention we go from Unreal -> Absolute and then git gives us repo relative so we just need to convert the absolute to also be repo relative and then we can compare them

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2532,6 +2535,80 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+	TArray<FString> ErrorMessages;
+	TArray<FString> DiffParametersLog{ TEXT("--pretty="), TEXT("--name-only"), FString::Printf(TEXT("HEAD..%s"), *BranchName), TEXT(""), TEXT("--") };
+	const bool bResultDiff = RunCommand(TEXT("diff"), InPathToGitBinary, InRepositoryRoot, DiffParametersLog, FilesToSync, DiffResults, ErrorMessages);
+	FilesToSync.RemoveAllSwap([&DiffResults](const FString& File)
```
This seems to always result in an empty list

Damon3000s replied:
By fixing the paths so we can compare, this has resolved the issue.

Damon3000s committed:
Resave the GitSourceControlUtils file

Applies our current editor config which strips whitespace

Damon3000s committed:
First pass at fixing Revert To Status Branch

The issue is that paths are in different formats and comparisons don't work (see comments). This works for the base game but requires more testing for plugin content. Matt has also recommended to look into "CreateStandardFilename" and "MakeStandardFilename" as a better alternative, but I want to get this "working" version in first.

Damon3000s committed:
Improve naming and how we substring

Damon3000s committed:
Revert "Resave the GitSourceControlUtils file"

This reverts commit c62f95fc1127ddb8817607356d78a5c1c58eb7c9.

Damon3000s committed:
Merge branch 'dev-3k' into revert_to_status_branch_clean

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2580,6 +2583,95 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+		// path and make it relative, but it will include the root folder (eg. The repo root name)
+		// which needs to be stripped for the comparison to work
+		FString RelativeFromRootInclusive = File;
+		FPaths::MakePathRelativeTo(RelativeFromRootInclusive, *InRepositoryRoot);
```
Suggestion: Is it better to convert the DiffResults to absolute instead of converting FilesToSync to relative?

Damon3000s replied:
Good suggestion. Made everything much simplier.

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2580,6 +2583,95 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+		FPaths::MakePathRelativeTo(RelativeFromRootInclusive, *InRepositoryRoot);
+
+		FString RelativeFromRoot;
+		RelativeFromRootInclusive.Split("/", nullptr, &RelativeFromRoot);
```
What happens if there is no forward slash?

Damon3000s replied:
Look at FPaths::NormalizeFilename

Damon3000s replied:
The repo relative paths are now turned into absolute paths for easier comparison

Damon3000s committed:
Simplify path conversions

Converting DiffResults to absolute paths is easier than converting absolute paths to git repo relative paths. Less code, less problems.

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2580,6 +2583,88 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+		// Git checkout is pulling content
+		TArray<FString> Results;
+		TArray<FString> Errors;
+		RunCommand("checkout", InPathToGitBinary, InRepositoryRoot, { BranchName, TEXT("--") }, FilesToSync, Results, Errors);
```
Before this we need to run REVERT on all the files

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2580,6 +2583,88 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+		// Git checkout is pulling content
+		TArray<FString> Results;
+		TArray<FString> Errors;
+		RunCommand("checkout", InPathToGitBinary, InRepositoryRoot, { BranchName, TEXT("--") }, FilesToSync, Results, Errors);
```
After the files have been checked out we need to reload everything

Damon3000s replied:
I'm finding that the file is being modified on disk but is not actually being reflected in engine, even after reloading the file manually or restarting the editor (without saving)

Damon3000s replied:
USourceControlHelpers::ApplyOperationAndReloadPackages does the job.

Unlinks the file
Downloads the file
Reloads the file
Relinks the file

That was UE doesn't know the file changed underneath it and it should stomp the change.

Damon3000s committed:
Use ApplyOperationAndReloadPackages

This will cause the file to be unlinked, reloaded and relinked. This way Unreal won't know the file has changed and stomp it.

Damon3000s commented:
Closing as I accidentally merged dev-3k into this branch when I shouldn't have.

All feedback has been addressed and a new branch has been created from dev that does not merge in dev-3k.
